mpartition.1
'\" t
.TH mpartition 1 "02Dec18" mtools-4.0.22
.SH Name
mpartition - partition an MSDOS hard disk
'\" t
.de TQ
.br
.ns
.TP \\$1
..

.tr \(is'
.tr \(if`
.tr \(pd"

.SH Note\ of\ warning
This manpage has been automatically generated from mtools's texinfo
documentation, and may not be entirely accurate or complete.  See the
end of this man page for details.
.PP
.SH Description
.PP
The \fR\&\f(CWmpartition\fR command is used to create MS-DOS file systems as
partitions.  This is intended to be used on non-Linux systems,
i.e. systems where fdisk and easy access to SCSI devices are not
available.  This command only works on drives whose partition variable
is set.
.PP
.ft I
.nf
\&\fR\&\f(CWmpartition\fR \fR\&\f(CW-p\fR \fIdrive\fR
\&\fR\&\f(CWmpartition\fR \fR\&\f(CW-r\fR \fIdrive\fR
\&\fR\&\f(CWmpartition\fR \fR\&\f(CW-I\fR [\fR\&\f(CW-B\fR \fIbootSector\fR] \fIdrive\fR 
\&\fR\&\f(CWmpartition\fR \fR\&\f(CW-a\fR \fIdrive\fR
\&\fR\&\f(CWmpartition\fR \fR\&\f(CW-d\fR \fIdrive\fR
\&\fR\&\f(CWmpartition\fR \fR\&\f(CW-c\fR [\fR\&\f(CW-s\fR \fIsectors\fR] [\fR\&\f(CW-h\fR \fIheads\fR]
[\fR\&\f(CW-t\fR \fIcylinders\fR] [\fR\&\f(CW-v\fR [\fR\&\f(CW-T\fR \fItype\fR] [\fR\&\f(CW-b\fR
\&\fIbegin\fR] [\fR\&\f(CW-l\fR length] [\fR\&\f(CW-f\fR]
\&\&
.fi
.ft R
 
.PP
Mpartition supports the following operations:
.TP
\&\fR\&\f(CWp\fR\ 
Prints a command line to recreate the partition for the drive.  Nothing
is printed if the partition for the drive is not defined, or an
inconsistency has been detected.  If verbose (\fR\&\f(CW-v\fR) is also set,
prints the current partition table.
.TP
\&\fR\&\f(CWr\fR\ 
Removes the partition described by \fIdrive\fR.
.TP
\&\fR\&\f(CWI\fR\ 
Initializes the partition table, and removes all partitions.
.TP
\&\fR\&\f(CWc\fR\ 
Creates the partition described by \fIdrive\fR.
.TP
\&\fR\&\f(CWa\fR\ 
"Activates" the partition, i.e. makes it bootable.  Only one partition
can be bootable at a time.
.TP
\&\fR\&\f(CWd\fR\ 
"Deactivates" the partition, i.e. makes it unbootable.
.PP
If no operation is given, the current settings are printed.
.PP
For partition creations, the following options are available:
.TP
\&\fR\&\f(CWs\ \fIsectors\fR\&\f(CW\fR\ 
The number of sectors per track of the partition (which is also the
number of sectors per track for the whole drive).
.TP
\&\fR\&\f(CWh\ \fIheads\fR\&\f(CW\fR\ 
The number of heads of the partition (which is also the number of heads
for the whole drive).  By default, the geometry information (number of
sectors and heads) is figured out from neighboring partition table
entries, or guessed from the size.
.TP
\&\fR\&\f(CWt\ \fIcylinders\fR\&\f(CW\fR\ 
The number of cylinders of the partition (not the number of cylinders of
the whole drive.
.TP
\&\fR\&\f(CWb\ \fIbegin\fR\&\f(CW\fR\ 
The starting offset of the partition, expressed in sectors. If begin is
not given, mpartition lets the partition begin at the start of the disk
(partition number 1), or immediately after the end of the previous
partition.
.TP
\&\fR\&\f(CWl\ \fIlength\fR\&\f(CW\fR\ 
The size (length) of the partition, expressed in sectors.  If end is not
given, mpartition figures out the size from the number of sectors, heads
and cylinders.  If these are not given either, it gives the partition
the biggest possible size, considering disk size and start of the next
partition.
.PP
The following option is available for all operation which modify the
partition table:
.TP
\&\fR\&\f(CWf\fR\ 
Usually, before writing back any changes to the partition, mpartition
performs certain consistency checks, such as checking for overlaps and
proper alignment of the partitions.  If any of these checks fails, the
partition table is not changes.  The \fR\&\f(CW-f\fR allows you to override
these safeguards.
.PP
The following options are available for all operations:
.TP
\&\fR\&\f(CWv\fR\ 
Together with \fR\&\f(CW-p\fR prints the partition table as it is now (no
change operation), or as it is after it is modified.
.TP
\&\fR\&\f(CWvv\fR\ 
If the verbosity flag is given twice, mpartition will print out a
hexdump of the partition table when reading it from and writing it to
the device.
.PP
The following option is available for partition table initialization:
.TP
\&\fR\&\f(CWB\ \fIbootSector\fR\&\f(CW\fR\ 
Reads the template master boot record from file \fIbootSector\fR.
.PP
.SH Choice\ of\ partition\ type
.PP
Mpartition proceeds as follows to pick a type for the partition:
.TP
-\ \ 
FAT32 partitions are assigned type 0x0C (``\fR\&\f(CWWin95 FAT32, LBA\fR'')
.TP
-\ \ 
For all others, if the partition fits entirely within the 65536 sector of the disk, assign 0x01 (``\fR\&\f(CWDOS FAT12, CHS\fR'') for FAT12 partition and 0x04 (``\fR\&\f(CWDOS FAT16, CHS\fR'') for FAT16 partitions
.TP
-\ \ 
If not covered by the above, assign 0x06 (``\fR\&\f(CWDOS BIG FAT16 CHS\fR'') if partition fits entirely within the first 1024 cylinders (CHS mode)
.TP
-\ \ 
All remaining cases get 0x0E (``\fR\&\f(CWWin95 BIG FAT16, LBA\fR'')
.PP
If number of fat bits is not known (not specified in drive's
definition), then FAT12 is assumed for all drives with less than 4096
sectors, and FAT16 for those with more than 4096 sectors.
.PP
This corresponds more or less to the definitions outlined at \fR\&\f(CWhttps://en.wikipedia.org/wiki/Partition_type#List_of_partition_IDs\fR
and
\&\fR\&\f(CWhttps://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-2000-server/cc977219(v=technet.10)\fR,
with two notable differences:
.TP
-\ \ 
If fat bits are unknown, the reference documents consider drives with
less than 32680 sectors to be FAT12. Mtools uses 4096 sectors as the
cutoff point, as older versions of DOS only support FAT12 on disks
with less than 4096 sectors (and these older versions are the ones
which would be most likely to use FAT12 in the first place).
.TP
-\ \ 
The reference documents use a 8GB (wikipedia) or a 4GB (Microsoft)
cutoff between 0x06 (\fR\&\f(CWDOS BIG FAT16 CHS\fR) and 0x0E. Mtools uses
1024 cylinders. This is because any partition beyond 1024 cylinders
must be LBA and cannot be CHS. 8GB works out to be the biggest
capacity which can be represented as CHS (63 sectors, 255 heads and
1024 cylinders). 4GB is the capacity limit for windows 2000, so it
makes sense that a documentation for windows 2000 would specify this
as the upper limit for any partition type.
.PP
.SH See\ Also
Mtools' texinfo doc
.SH Viewing\ the\ texi\ doc
This manpage has been automatically generated from mtools's texinfo
documentation. However, this process is only approximative, and some
items, such as crossreferences, footnotes and indices are lost in this
translation process.  Indeed, these items have no appropriate
representation in the manpage format.  Moreover, not all information has
been translated into the manpage version.  Thus I strongly advise you to
use the original texinfo doc.  See the end of this manpage for
instructions how to view the texinfo doc.
.TP
* \ \ 
To generate a printable copy from the texinfo doc, run the following
commands:
 
.nf
.ft 3
.in +0.3i
    ./configure; make dvi; dvips mtools.dvi
.fi
.in -0.3i
.ft R
.PP
 
\&\fR
.TP
* \ \ 
To generate a html copy,  run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make html
.fi
.in -0.3i
.ft R
.PP
 
\&\fRA premade html can be found at
\&\fR\&\f(CW\(ifhttp://www.gnu.org/software/mtools/manual/mtools.html\(is\fR
.TP
* \ \ 
To generate an info copy (browsable using emacs' info mode), run:
 
.nf

NEWS
v4_0_22
	- Fixed -f flag for mformat (size is KBytes, rather than sectors)
	- Fixed toupper/tolower usage (unsigned char rather than plain signed)
v4_0_21
	- Fixed compilation for MingW
	- After MingW compilation, make sure executable has .exe extension
	- Addressed compiler warnings
	- Fixed length handling in character set conversion (Unicode file names)
	- Fixed matching of character range, when containing Unicode
	characters (mdir "c:test[α-ω].exe")
	- Fixed initialization of my_scsi_cmd constructor
v4_0_20
	- initialize directory entries to 0
	- bad message "Too few sectors" replaced with "Too many sectors"
	- apostrophe in mlabel no longer causes generation of long entry
	- option to fake system date for file creation using the
	SOURCE_DATE_EPOCH environment variables

	- can now be compiled with "clang" compiler
	- fallback function for strndup, for those platforms that don't have it
	- fixed a number of -Wextra warnings

	- new compressed archive formats for uz/lz

	- allow to specify number of reserved sectors for FAT32.
	- file/device locking with timeout (rather than immediate failure)
	- fixed support for BPB-less legacy formats.
	- removed check that disk must be an integer number of tracks.

	- removed .eh/.oh macros from manual pages
v4_0_19
	- Fix for short file names starting with character 0xE5
	(by remapping it to 0x5)
	- mpartition: Partition types closer to what Microsoft uses
	- mformat: figure out LBA geometry as last resort if geometry
	is neither specified in config and/or commandline, nor can be
	queried from the device
	- mformat: use same default cluster size by size as Microsoft
	for FAT32
	- additional sanity checks
	- document how cluster size is picked in mformat.c man page
	- document how partition types are picked in mpartition.c man page
v4_0_18	
	Fix for names of iconv encodings on AIX
	Fix mt_size_t on NetBSD
	Fixed compilation on Mingw
	Fixed doc (especially mformat)
	Fix mformat'ing of FAT12 filesystems with huge cluster sizes
	Minfo prints image file name in mformat command line if an image
	file name was given
	Always generate gzip-compressed RPMs, in order to remain
	compatible with older distributions
	Fixed buffer overflow with drive letter in mclasserase
v4_0_17
	mbadblocks now takes a list of bad blocks (either as sectors
	or as clusters)
	mbadblocks now is able to do write scanning for bad blocks
	mshowfat can show cluster of specific offset
	Enable mtools to deal with very small sector sizes...
	Fixed encoding of all-lowercase names (no need to mangle
	these)
	Consider every directory entry after an ENDMARK (0x00) to be deleted
	After writing a new entry at end of a directory, be sure to also add
	an ENDMARK (0x00)

	Deal with possibility of a NULL pointer being returned by
	localtime during timestamp conversion
v4_0_16
	configure.in fixes
	fixed formatting of fat_size_calculation.tex document
	compatibility with current autoconfig versions
	Make it clear that label is limited to 11 characters
	Fixed typo in initialization of FAT32 info sector
v4_0_15
	Added missing -i option to mshortname
	Splitted .deb package into mtools and floppyd in order to
	match Ubuntu
v4_0_14
	New mshortname command
	Fix floppyd for disks bigger than 2 Gig
	Remove obsolete -z flag
	Remove now unsupported AC_USE_SYSTEM_EXTENSIONS
	Fixed output formatting of mdir if MTOOLS_DOTTED_DIR is set
	Mformat now correctly writes backup boot sector
	Fixed signedness of serial number in mlabel
	Fixed buffer size problem in mlabel
	Make mlabel write backup boot sector if FAT32
	Catch situation where both clear and new label are given to mlabel
	Quote filename parameters to scripts
	Mformat: Close file descriptor for boot sector

mpartition.c
/*  Copyright 1997-2003,2005-2007,2009 Alain Knaff.
 *  This file is part of mtools.
 *
 *  Mtools is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Mtools is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Mtools.  If not, see <http://www.gnu.org/licenses/>.
 *
 * mformat.c
 */
#define DONT_NEED_WAIT

#include "sysincludes.h"
#include "msdos.h"
#include "mtools.h"
#include "mainloop.h"
#include "fsP.h"
#include "file.h"
#include "plain_io.h"
#include "nameclash.h"
#include "buffer.h"
#include "scsi.h"
#include "partition.h"

#ifdef OS_linux
#include "linux/hdreg.h"

#define _LINUX_STRING_H_
#define kdev_t int
#include "linux/fs.h"
#undef _LINUX_STRING_H_

#endif

#define tolinear(x) \
(sector(x)-1+(head(x)+cyl(x)*used_dev->heads)*used_dev->sectors)


static __inline__ void print_hsc(hsc *h)
{
	printf(" h=%d s=%d c=%d\n",
	       head(*h), sector(*h), cyl(*h));
}

static void set_offset(hsc *h, int offset, int heads, int sectors)
{
	int head, sector, cyl;

	if(! heads || !sectors)
		head = sector = cyl = 0; /* linear mode */
	else {
		sector = offset % sectors;
		offset = offset / sectors;

		head = offset % heads;
		cyl = offset / heads;
		if(cyl > 1023) cyl = 1023;
	}

	h->head = head;
	h->sector = ((sector+1) & 0x3f) | ((cyl & 0x300)>>2);
	h->cyl = cyl & 0xff;
}

void setBeginEnd(struct partition *partTable,
		 unsigned int begin, unsigned int end,
		 unsigned int heads, unsigned int sectors,
		 int activate, int type, int fat_bits)
{
	set_offset(&partTable->start, begin, heads, sectors);
	set_offset(&partTable->end, end-1, heads, sectors);
	set_dword(partTable->start_sect, begin);
	set_dword(partTable->nr_sects, end-begin);
	if(activate)
		partTable->boot_ind = 0x80;
	else
		partTable->boot_ind = 0;
	if(!type) {
		if (fat_bits == 0) {
			/**
			 * Fat bits unknown / not specified. We look
			 * at size to get a rough estimate what FAT
			 * bits are used.  Note: this is only an
			 * estimate, the precise calculation would
			 * involve the number of clusters, which is
			 * not necessarily known here.
			 */
			/* cc977219 would have a cutoff number of 32680,
			 * corresponding to a FAT12 partition with 4K
			 * clusters, however other information hints that
			 * only partitions with less than 4096 sectors are
			 * considered */
			if(end-begin < 4096)
				fat_bits = 12;
			else
				fat_bits = 16;
		}

		/* Description of various partition types in
		 * https://en.wikipedia.org/wiki/Partition_type#List_of_partition_IDs
		 * and
		 * https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-2000-server/cc977219(v=technet.10)
		 */
		if (fat_bits == 32)
			/* FAT 32 partition. For now, we disregard the
			 * possibility of FAT 32 CHS partitions */
			type = 0x0C; /* Win95 FAT32, LBA */
		else if (end < 65536) {
			/* FAT 12 or FAT 16 partitions which fit entirely below
			   the 32M mark */
			/* The 32M restriction doesn't apply to logical
			   partitions within an extended partition, but for the
			   moment mpartition only makes primary partitions */
			if(fat_bits == 12)
				/* FAT 12 partition */
				type = 0x01; /* DOS FAT12, CHS */
			else if (fat_bits == 16)
				/* FAT 16 partition */
				type = 0x04; /* DOS FAT16, CHS */
		} else if (end <  sectors * heads * 1024)
			/* FAT 12 or FAT16 partition above the 32M
			 * mark but below the 1024 cylinder mark.
			 * Indeed, there can be no CHS partition
			 * beyond 1024 cylinders */
			type = 0x06; /* DOS BIG FAT16 or FAT12, CHS */
		else
			type = 0x0E; /* Win95 BIG FAT16, LBA */
	}
	partTable->sys_ind = type;
}

int consistencyCheck(struct partition *partTable, int doprint, int verbose,
		     int *has_activated, unsigned int *last_end,
		     unsigned int *j,
		     struct device *used_dev, int target_partition)
{
	int i;
	unsigned int inconsistency;

	*j = 0;
	*last_end = 1;

	/* quick consistency check */
	inconsistency = 0;
	*has_activated = 0;
	for(i=1; i<5; i++){
		if(!partTable[i].sys_ind)
			continue;
		if(partTable[i].boot_ind)
			(*has_activated)++;
		if((used_dev &&
		    (used_dev->heads != head(partTable[i].end)+1 ||
		     used_dev->sectors != sector(partTable[i].end))) ||
		   sector(partTable[i].start) != 1){
			fprintf(stderr,
				"Partition %d is not aligned\n",
				i);
			inconsistency=1;
		}

		if(*j &&
		   *last_end > BEGIN(partTable[i])) {
			fprintf(stderr,
				"Partitions %d and %d badly ordered or overlapping\n",
				*j,i);
			inconsistency=1;
		}

		*last_end = END(partTable[i]);
		*j = i;

		if(used_dev &&
		   cyl(partTable[i].start) != 1023 &&
		   tolinear(partTable[i].start) != BEGIN(partTable[i])) {
			fprintf(stderr,
				"Start position mismatch for partition %d\n",
				i);
			inconsistency=1;
		}
		if(used_dev &&
		   cyl(partTable[i].end) != 1023 &&
		   tolinear(partTable[i].end)+1 != END(partTable[i])) {
			fprintf(stderr,
				"End position mismatch for partition %d\n",
				i);
			inconsistency=1;
		}

		if(doprint && verbose) {
			if(i==target_partition)
				putchar('*');
			else
				putchar(' ');
			printf("Partition %d\n",i);

			printf("  active=%x\n", partTable[i].boot_ind);
			printf("  start:");
			print_hsc(&partTable[i].start);
			printf("  type=0x%x\n", partTable[i].sys_ind);
			printf("  end:");
			print_hsc(&partTable[i].end);
			printf("  start=%d\n", BEGIN(partTable[i]));
			printf("  nr=%d\n", _DWORD(partTable[i].nr_sects));
			printf("\n");
		}
	}
	return inconsistency;
}

/* setsize function.  Determines scsicam mapping if this cannot be inferred from
 * any existing partitions. Shamelessly snarfed from the Linux kernel ;-) */

/*
 * Function : static int setsize(unsigned long capacity,unsigned int *cyls,
 *	unsigned int *hds, unsigned int *secs);
 *
 * Purpose : to determine a near-optimal int 0x13 mapping for a
 *	SCSI disk in terms of lost space of size capacity, storing
 *	the results in *cyls, *hds, and *secs.
 *
 * Returns : -1 on failure, 0 on success.
 *
 * Extracted from
 *
 * WORKING                                                    X3T9.2
 * DRAFT                                                        792D
 *
 *
 *                                                        Revision 6
 *                                                         10-MAR-94
 * Information technology -
 * SCSI-2 Common access method
 * transport and SCSI interface module
 *
 * ANNEX A :
 *
 * setsize() converts a read capacity value to int 13h
 * head-cylinder-sector requirements. It minimizes the value for
 * number of heads and maximizes the number of cylinders. This
 * will support rather large disks before the number of heads
 * will not fit in 4 bits (or 6 bits). This algorithm also
 * minimizes the number of sectors that will be unused at the end
 * of the disk while allowing for very large disks to be
 * accommodated. This algorithm does not use physical geometry.
 */

static int setsize(unsigned long capacity,unsigned int *cyls,unsigned int *hds,
    unsigned int *secs) {
    unsigned int rv = 0;
    unsigned long heads, sectors, cylinders, temp;

    cylinders = 1024L;			/* Set number of cylinders to max */
    sectors = 62L;      		/* Maximize sectors per track */

    temp = cylinders * sectors;		/* Compute divisor for heads */
    heads = capacity / temp;		/* Compute value for number of heads */
    if (capacity % temp) {		/* If no remainder, done! */
    	heads++;                	/* Else, increment number of heads */
    	temp = cylinders * heads;	/* Compute divisor for sectors */
    	sectors = capacity / temp;	/* Compute value for sectors per
					       track */
    	if (capacity % temp) {		/* If no remainder, done! */
      	    sectors++;                  /* Else, increment number of sectors */
      	    temp = heads * sectors;	/* Compute divisor for cylinders */
      	    cylinders = capacity / temp;/* Compute number of cylinders */
      	}
    }
    if (cylinders == 0) rv=(unsigned)-1;/* Give error if 0 cylinders */

    *cyls = (unsigned int) cylinders;	/* Stuff return values */
    *secs = (unsigned int) sectors;
    *hds  = (unsigned int) heads;
    return(rv);
}

static void setsize0(unsigned long capacity,unsigned int *cyls,
		     unsigned int *hds, unsigned int *secs)
{
	int r;

	/* 1. First try "Megabyte" sizes */
	if(capacity < 1024 * 2048 && !(capacity % 1024)) {
		*cyls = capacity >> 11;
		*hds  = 64;
		*secs = 32;
		return;
	}

	/* then try scsicam's size */
	r = setsize(capacity,cyls,hds,secs);
	if(r || *hds > 255 || *secs > 63) {
		/* scsicam failed. Do megabytes anyways */
		*cyls = capacity >> 11;
		*hds  = 64;
		*secs = 32;
		return;
	}
}


static void usage(int ret) NORETURN;
static void usage(int ret)
{
	fprintf(stderr,
		"Mtools version %s, dated %s\n", mversion, mdate);
	fprintf(stderr,
		"Usage: %s [-pradcv] [-I] [-B bootsect-template] [-s sectors] "
			"[-t cylinders] "
		"[-h heads] [-T type] [-b begin] [-l length] "
		"drive\n", progname);
	exit(ret);
}

void mpartition(int argc, char **argv, int dummy UNUSEDP) NORETURN;
void mpartition(int argc, char **argv, int dummy UNUSEDP)
{
	Stream_t *Stream;
	unsigned int dummy2;

	unsigned int i,j;

	int sec_per_cyl;
	int doprint = 0;
	int verbose = 0;
	int create = 0;
	int force = 0;
	unsigned int length = 0;
	int do_remove = 0;
	int initialize = 0;
	unsigned int tot_sectors=0;
	int type = 0;
	int begin_set = 0;
	int size_set = 0;
	int end_set = 0;
	unsigned int last_end = 0;
	int activate = 0;
	int has_activated = 0;
	int inconsistency=0;
	unsigned int begin=0;
	unsigned int end=0;
	int sizetest=0;
	int dirty = 0;
	int open2flags = NO_OFFSET;

	int c;
	struct device used_dev;
	int argtracks, argheads, argsectors;

	char drive, name[EXPAND_BUF];
	unsigned char buf[512];
	struct partition *partTable=(struct partition *)(buf+ 0x1ae);
	struct device *dev;
	char errmsg[200];
	char *bootSector=0;

	argtracks = 0;
	argheads = 0;
	argsectors = 0;

	/* get command line options */
	if(helpFlag(argc, argv))
		usage(0);
	while ((c = getopt(argc, argv, "i:adprcIT:t:h:s:fvpb:l:S:B:")) != EOF) {
		switch (c) {
			case 'i':
				set_cmd_line_image(optarg);
				break;
			case 'B':
				bootSector = optarg;
				break;
			case 'a':
				/* no privs, as it could be abused to
				 * make other partitions unbootable, or
				 * to boot a rogue kernel from this one */
				open2flags |= NO_PRIV;
				activate = 1;
				dirty = 1;
				break;
			case 'd':
				activate = -1;
				dirty = 1;
				break;
			case 'p':
				doprint = 1;
				break;
			case 'r':
				do_remove = 1;
				dirty = 1;
				break;
			case 'I':
				/* could be abused to nuke all other
				 * partitions */
				open2flags |= NO_PRIV;
				initialize = 1;
				dirty = 1;
				break;
			case 'c':
				create = 1;
				dirty = 1;
				break;

			case 'T':
				/* could be abused to "manually" create
				 * extended partitions */
				open2flags |= NO_PRIV;
				type = strtoul(optarg,0,0);
				break;

			case 't':
				argtracks = atoi(optarg);
				break;
			case 'h':
				argheads = atoi(optarg);
				break;
			case 's':
				argsectors = atoi(optarg);
				break;

			case 'f':
				/* could be abused by creating overlapping
				 * partitions and other such Snafu */
				open2flags |= NO_PRIV;
				force = 1;
				break;

			case 'v':
				verbose++;
				break;
			case 'S':
				/* testing only */
				/* could be abused to create partitions
				 * extending beyond the actual size of the
				 * device */
				open2flags |= NO_PRIV;
				tot_sectors = strtoul(optarg,0,0);
				sizetest = 1;
				break;
			case 'b':
				begin_set = 1;
				begin = strtoul(optarg, NULL, 10);
				break;
			case 'l':
				size_set = 1;
				length = strtoul(optarg, NULL, 10);
				break;

			default:
				usage(1);
		}
	}

	if (argc - optind != 1 ||
	    !argv[optind][0] || argv[optind][1] != ':')
		usage(1);

	drive = ch_toupper(argv[optind][0]);

	/* check out a drive whose letter and parameters match */
	sprintf(errmsg, "Drive '%c:' not supported", drive);
	Stream = 0;
	for(dev=devices;dev->drive;dev++) {
		int mode ;

		FREE(&(Stream));
		/* drive letter */
		if (dev->drive != drive)
			continue;
		if (dev->partition < 1 || dev->partition > 4) {
			sprintf(errmsg,
				"Drive '%c:' is not a partition",
				drive);
			continue;
		}
		used_dev = *dev;

		SET_INT(used_dev.tracks, argtracks);
		SET_INT(used_dev.heads, argheads);
		SET_INT(used_dev.sectors, argsectors);

		expand(dev->name, name);

		mode = dirty ? O_RDWR : O_RDONLY;
		if(initialize)
 			mode |= O_CREAT;

#ifdef USING_NEW_VOLD
		strcpy(name, getVoldName(dev, name));
#endif
		Stream = SimpleFileOpen(&used_dev, dev, name, mode,
					errmsg, open2flags, 1, 0);

		if (!Stream) {
#ifdef HAVE_SNPRINTF
			snprintf(errmsg,199,"init: open: %s", strerror(errno));
#else
			sprintf(errmsg,"init: open: %s", strerror(errno));
#endif
			continue;
		}


		/* try to find out the size */
		if(!sizetest)
			tot_sectors = 0;
		if(IS_SCSI(dev)) {
			unsigned char cmd[10];
			unsigned char data[10];
			cmd[0] = SCSI_READ_CAPACITY;
			memset ((void *) &cmd[2], 0, 8);
			memset ((void *) &data[0], 137, 10);
			scsi_cmd(get_fd(Stream), cmd, 10, SCSI_IO_READ,
				 data, 10, get_extra_data(Stream));

			tot_sectors = 1 +
				(data[0] << 24) +
				(data[1] << 16) +
				(data[2] <<  8) +
				(data[3]      );
			if(verbose)
				printf("%d sectors in total\n", tot_sectors);
		}

#ifdef OS_linux
		if (tot_sectors == 0) {
			ioctl(get_fd(Stream), BLKGETSIZE, &tot_sectors);
		}
#endif

		/* read the partition table */
		if (READS(Stream, (char *) buf, 0, 512) != 512 && !initialize){
#ifdef HAVE_SNPRINTF
			snprintf(errmsg, 199,
				"Error reading from '%s', wrong parameters?",
				name);
#else
			sprintf(errmsg,
				"Error reading from '%s', wrong parameters?",
				name);
#endif
			continue;
		}
		if(verbose>=2)
			print_sector("Read sector", buf, 512);
		break;
	}

	/* print error msg if needed */
	if ( dev->drive == 0 ){
		FREE(&Stream);
		fprintf(stderr,"%s: %s\n", argv[0],errmsg);
		exit(1);
	}

	if((used_dev.sectors || used_dev.heads) &&
	   (!used_dev.sectors || !used_dev.heads)) {
		fprintf(stderr,"You should either indicate both the number of sectors and the number of heads,\n");
		fprintf(stderr," or none of them\n");
		exit(1);
	}

	if(initialize) {
		if (bootSector) {
			int fd;
			fd = open(bootSector, O_RDONLY | O_BINARY | O_LARGEFILE);
			if (fd < 0) {
				perror("open MBR");
				exit(1);
			}
			if(read(fd, (char *) buf, 512) < 512) {
				perror("read MBR");
				exit(1);
			}
		}
		memset((char *)(partTable+1), 0, 4*sizeof(*partTable));
		set_word(((unsigned char*)buf)+510, 0xaa55);
	}

	/* check for boot signature, and place it if needed */
	if((buf[510] != 0x55) || (buf[511] != 0xaa)) {
		fprintf(stderr,"Boot signature not set\n");
		fprintf(stderr,
			"Use the -I flag to initialize the partition table, and set the boot signature\n");
		inconsistency = 1;
	}

	if(do_remove){
		if(!partTable[dev->partition].sys_ind)
			fprintf(stderr,
				"Partition for drive %c: does not exist\n",
				drive);
		if((partTable[dev->partition].sys_ind & 0x3f) == 5) {
			fprintf(stderr,
				"Partition for drive %c: may be an extended partition\n",
				drive);
			fprintf(stderr,
				"Use the -f flag to remove it anyways\n");
			inconsistency = 1;
		}
		memset(&partTable[dev->partition], 0, sizeof(*partTable));
	}

	if(create && partTable[dev->partition].sys_ind) {
		fprintf(stderr,
			"Partition for drive %c: already exists\n", drive);
		fprintf(stderr,
			"Use the -r flag to remove it before attempting to recreate it\n");
	}


	/* find out number of heads and sectors, and whether there is
	* any activated partition */
	has_activated = 0;
	for(i=1; i<5; i++){
		if(!partTable[i].sys_ind)
			continue;

		if(partTable[i].boot_ind)
			has_activated++;

		/* set geometry from entry */
		if (!used_dev.heads)
			used_dev.heads = head(partTable[i].end)+1;
		if(!used_dev.sectors)
			used_dev.sectors = sector(partTable[i].end);
		if(i<dev->partition && !begin_set)
			begin = END(partTable[i]);
		if(i>dev->partition && !end_set && !size_set) {
			end = BEGIN(partTable[i]);
			end_set = 1;
		}
	}

#ifdef OS_linux
	if(!used_dev.sectors && !used_dev.heads) {
		if(!IS_SCSI(dev)) {
			struct hd_geometry geom;
			if(ioctl(get_fd(Stream), HDIO_GETGEO, &geom) == 0) {
				used_dev.heads = geom.heads;
				used_dev.sectors = geom.sectors;
			}
		}
	}
#endif

	if(!used_dev.sectors && !used_dev.heads) {
		if(tot_sectors)
			setsize0(tot_sectors,&dummy2,&used_dev.heads,
					 &used_dev.sectors);
		else {
			used_dev.heads = 64;
			used_dev.sectors = 32;
		}
	}

	if(verbose)
		fprintf(stderr,"sectors: %d heads: %d %d\n",
			used_dev.sectors, used_dev.heads, tot_sectors);

	sec_per_cyl = used_dev.sectors * used_dev.heads;
	if(create) {
		if(!end_set && tot_sectors) {
			end = tot_sectors - tot_sectors % sec_per_cyl;
			end_set = 1;
		}

		/* if the partition starts right at the beginning of
		 * the disk, keep one track unused to allow place for
		 * the master boot record */
		if(!begin && !begin_set)
			begin = used_dev.sectors;
		if(!size_set && used_dev.tracks) {
			size_set = 2;
			length = sec_per_cyl * used_dev.tracks;

			/*  round the size in order to take
			 * into account any "hidden" sectors */

			/* do we anchor this at the beginning ?*/
			if(begin_set || dev->partition <= 2 || !end_set)
				length -= begin % sec_per_cyl;
			else if(end - length < begin)
				/* truncate any overlap */
				length = end - begin;
		}
		if(size_set) {
			if(!begin_set && dev->partition >2 && end_set)
				begin = end - length;
			else
				end = begin + length;
		} else if(!end_set) {
			fprintf(stderr,"Unknown size\n");
			exit(1);
		}

		setBeginEnd(&partTable[dev->partition], begin, end,
			    used_dev.heads, used_dev.sectors,
			    !has_activated, type,
			    dev->fat_bits);
	}

	if(activate) {
		if(!partTable[dev->partition].sys_ind) {
			fprintf(stderr,
				"Partition for drive %c: does not exist\n",
				drive);
		} else {
			switch(activate) {
				case 1:
					partTable[dev->partition].boot_ind=0x80;
					break;
				case -1:
					partTable[dev->partition].boot_ind=0x00;
					break;
			}
		}
	}


	inconsistency |= consistencyCheck(partTable, doprint, verbose,
					  &has_activated, &last_end, &j,
					  &used_dev, dev->partition);

	if(doprint && !inconsistency && partTable[dev->partition].sys_ind) {
		printf("The following command will recreate the partition for drive %c:\n",
		       drive);
		used_dev.tracks =
			(_DWORD(partTable[dev->partition].nr_sects) +
			 (BEGIN(partTable[dev->partition]) % sec_per_cyl)) /
			sec_per_cyl;
		printf("mpartition -c -t %d -h %d -s %d -b %u %c:\n",
		       used_dev.tracks, used_dev.heads, used_dev.sectors,
		       BEGIN(partTable[dev->partition]), drive);
	}

	if(tot_sectors && last_end >tot_sectors) {
		fprintf(stderr,
			"Partition %d exceeds beyond end of disk\n",
			j);
		exit(1);
	}


	switch(has_activated) {
		case 0:
			fprintf(stderr,
				"Warning: no active (bootable) partition present\n");
			break;
		case 1:
			break;
		default:
			fprintf(stderr,
				"Warning: %d active (bootable) partitions present\n",
				has_activated);
			fprintf(stderr,
				"Usually, a disk should have exactly one active partition\n");
			break;
	}

	if(inconsistency && !force) {
		fprintf(stderr,
			"inconsistency detected!\n" );
		if(dirty)
			fprintf(stderr,
				"Retry with the -f switch to go ahead anyways\n");
		exit(1);
	}

	if(dirty) {
		/* write data back to the disk */
		if(verbose>=2)
			print_sector("Writing sector", buf, 512);
		if (WRITES(Stream, (char *) buf, 0, 512) != 512) {
			fprintf(stderr,"Error writing partition table");
			exit(1);
		}
		if(verbose>=3)
			print_sector("Sector written", buf, 512);
	}
	FREE(&Stream);
	exit(0);
}
